### DISCUSSION TITLE
flax equivalent of `register_buffer` from pytorch

### DISCUSSION THREAD

**Discussion by ysig** [google/flax#2381] [Q&A] [answered]
Hi,

I'm searching a way to write the equivalent of the following really silly Pytorch module in Flax but I haven't found a way to do it. The important thing is that the constant should be loadable and saveable upon checkpoing.

```
    class SillyModule(nn.Module):
        def __init__(self, ):
            super().__init__()
            self.register_buffer('constant', torch.randn(1, 128))
    
        def forward(self, x):
            return torch.matmul(x, self.B)
```

Does anybody know how to do this? What is the equivalent of `register_buffer` in flax?

**Comment by marcvanzee** [comment, marked as answer]
You can store this as a separate variable collection, which you can treat separately from trainable parameters. Here is an example containing both trainable and non-trainable parameters:

```py
from flax import linen as nn
from flax.linen.initializers import uniform
from jax import random
import jax.numpy as jnp

class SillyModule(nn.Module):
  hidden_size: int = 10
  out_features: int = 128

  @nn.compact
  def __call__(self, x):
    x = nn.Dense(self.hidden_size)(x)
    rng = self.make_rng('consts')
    a = self.variable('constants', 'a', uniform(), rng, (10, self.out_features))
    return x @ a.value

rngs = {'params': random.PRNGKey(0), 'consts': random.PRNGKey(1) }
print(SillyModule().tabulate(rngs, jnp.zeros((3, 4))))
```

```

                             SillyModule Summary                             
┏━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━┳━━━━━━━━━━━━━━━━━━━━━━━┓
┃ path        ┃ outputs        ┃ constants          ┃ params                ┃
┡━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━╇━━━━━━━━━━━━━━━━━━━━━━━┩
│ Inputs      │ float32[3,4]   │                    │                       │
├─────────────┼────────────────┼────────────────────┼───────────────────────┤
│ Dense_0     │ float32[3,10]  │                    │ bias: float32[10]     │
│             │                │                    │ kernel: float32[4,10] │
│             │                │                    │                       │
│             │                │                    │ 50 (200 B)            │
├─────────────┼────────────────┼────────────────────┼───────────────────────┤
│ SillyModule │ float32[3,128] │ a: float32[10,128] │                       │
│             │                │                    │                       │
│             │                │ 1,280 (5.1 KB)     │                       │
├─────────────┼────────────────┼────────────────────┼───────────────────────┤
│             │          Total │ 1,280 (5.1 KB)     │ 50 (200 B)            │
└─────────────┴────────────────┴────────────────────┴───────────────────────┘
                                                                             
                      Total Parameters: 1,330 (5.3 KB)                   
```

A few things to note:
* You should pass rngs for both params and consts when you are initializing, since both variable collections need an rng. The `params` rng is used to initialize the Dense (which happens under the hood in the `Dense`), while the `consts` rng is used to initialize the constant (which happens explicitly when calling `self.variable`).
* When you train your module, you will only get the gradients with respect to the `params` variable collection and you pass in the `consts` manually. For an example of how to do this, see our [imagenet example](https://github.com/google/flax/blob/f75454111ce2a12eee196d31fa64ee37e2be9509/examples/imagenet/train.py#L108).

**Reply by ysig** [reply]
Hi & thank you for your reply!

I see you define the variable at __call__ time. How can I do the same during setup?
Also I don't see in the `imagenet` example where `consts` is passed?
(also isn't there a way to not pass things by hand? I find this to be complicated for a modern deep learning library.)

**Reply by cgarciae** [reply]
Hey @ysig, you can also define it inside setup:

```python
class SillyModule(nn.Module):
  hidden_size: int = 10
  out_features: int = 128

  def setup(self):
    rng = self.make_rng('consts')
    self.a = self.variable('constants', 'a', uniform(), rng, (10, self.out_features))

  @nn.compact
  def __call__(self, x):
    x = nn.Dense(self.hidden_size)(x)
    return x @ self.a.value
```
